JEP: Type Annotations (#11859)
jax-ml/jax: docs/jep/11859-type-annotations.md

# Type Annotation Roadmap for JAXS
_jakevdp@_
_Aug 2022_

> ## Background

Reviewer: Nit: Maybe add spacing after `#`/`##` section titles here and across the doc (ref: https://google.github.io/styleguide/docguide/style.html#add-spacing-to-headings)

For example:

```suggestion
## Background

```

To some degree, type annotations and static type checking have become an integral part of many Python development workflows, and to this end we have added annotations in a number of places throughout the JAX API.
The current state of type annotations in JAX is a bit patchwork, and efforts to add more have been hampered by more fundamental design questions.
This doc attempts to summarize those issues and generate a roadmap for the goals and non-goals of type annotations in JAX.

Why do we need such a roadmap? Better/more comprehensive type annotations are a frequent request from users, both internally and externally.
In addition, we frequently receive pull requests from external users seeking to improve JAX's type annotations: it's not always clear to the JAX team member reviewing the code whether such contributions are beneficial, particularly when they introduce complex Protocols to address the challenges inherent to full-fledged annotation of JAX's use of Python.

> This document details JAX's goals and recommendations for type annotations within the package.

Reviewer: General comment: Great TL;DR 👍

## Why Type Annotations?

> There are a number of reasons that a Python project might wish to annotate their code-base; I'll summarize them in this document as Level 1, Level 2, and Level 3.

Reviewer: Nit: Maybe "I'll" -> "we'll", since across the doc you say "we" (e.g. "we may", "we could", "we have", etc), even though there's only one author.

```suggestion
There are a number of reasons that a Python project might wish to annotate their code-base. We'll summarize them in this document as Level 1, Level 2, and Level 3.
```

### Level 1: Annotations as Documentation
When originally introduced in [PEP 3107](https://peps.python.org/pep-3107/), type annotations were motivated partly by the ability to use them as concise, inline documentation of function parameter types and return types. JAX has long utilized annotations in this manner; an example is the common pattern of creating type names aliased to Any. An example can be found in `lax/slicing.py` [[source](http://google3/third_party/py/jax/_src/lax/slicing.py;l=48-54;rcl=446366018)]:
```python
Array = Any
Shape = core.Shape

def slice(operand: Array, start_indices: Sequence[int],
          limit_indices: Sequence[int],
          strides: Optional[Sequence[int]] = None) -> Array:
  ...
```
For the purposes of static type checking, this use of `Array = Any` for array annotations puts no constraint on the argument values (`Any` is equivalent to no annotation at all) but it does serve as a form of useful in-code documentation for the developer.

> For the sake of generated documentation, the name of the alias gets lost (the [HTML docs](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.slice.html) for `jax.lax.slice` report operand as type Any), so the documentation benefit does not go beyond the source code.

Reviewer: Nit: Use code font for `Any`

```suggestion
For the sake of generated documentation, the name of the alias gets lost (the [HTML docs](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.slice.html) for `jax.lax.slice` report operand as type `Any`), so the documentation benefit does not go beyond the source code.
```

This use of type checking requires going further than the simple aliases used above; for example, knowing that the `slice` function returns an alias of `Any` named `Array` does not add any useful information to the code completion engine. However, were we to annotate the function with a `DeviceArray` return type, the autocomplete would know how to populate the namespace of the result, and thus be able to suggest more relevant autocompletions during the course of development.

JAX has begun to add this level of type annotation in a few places; one example is the `jnp.ndarray` return type within the `jax.random` package [[source](http://google3/third_party/py/jax/_src/random.py;l=359;rcl=445522278)]:

> ```python

Reviewer: Nit: Add new lines before and after this code block (ref: https://google.github.io/styleguide/docguide/style.html#add-spacing-to-headings)

Full static type checking is the strictest of all the type annotation applications, because it will surface an error any time your type annotations are not precisely correct.
On the one hand, this is nice because your static type analysis may catch faulty type annotations (for example, a case where a `DeviceArray` method is missing from the `jnp.ndarray` abstract class).

On the other hand, this strictness can make the type checking process very brittle in packages that often rely on duck-typing rather than strict type-safe APIs.
You'll currently find code comments like `#type: ignore` (for mypy) or `#pytype: disable` (for pytype) peppered throughout the JAX codebase in several hundred places.
These typically represent cases where typing problems have arisen; they may be inaccuracies in JAX type annotations, or inaccuracies in the static type checker's ability to correctly follow the control flow in the code.
On occasion, they are due to real & subtle bugs in the behavior of pytype and/or mypy.
In rare cases, they may be due to the fact that JAX uses Python patterns that are difficult or even impossible to express in terms of Python's static type annotation syntax.

> ## Type Annotation Challenges for JAX

Reviewer: Nit: May be best to use sentence case. Because you use lower-case for `pytype` and `mype`, capitalizing some words may look strange. Applies here and across the entire doc. Ref: https://developers.google.com/style/capitalization?hl=en#capitalization-in-titles-and-headings

Example:

```suggestion
## Type annotation challenges for JAX
```

Reviewer: @jakevdp Forgot to mention

> "... capitalize only the first word in the title, the first word in a subheading after a colon, and any proper nouns or other terms that are always capitalized a certain way."

From https://developers.google.com/style/capitalization#capitalization-in-titles-and-headings

So:

```diff
- ## Level 1: type annotation is cool
+ ## Level 1: Type annotation is cool
```

> As we've seen above, JAX currently has type annotations that are a mish-mash of styles, and aimed at all three areas of type annotation application above.

Reviewer: Nit:

- For users who aren't very proficient in English, maybe it's best to avoid using words like "mish-mash" (ref: Global audience https://developers.google.com/style/translation?hl=en#be-inclusive)
- "Seen" may not be 100% a11y-friendly. How about "demonstrated"?

Example:

```suggestion
As demonstrated above, JAX currently has type annotations that are a mixture of different styles, and aimed at all three areas of type annotation application above.
```

🤷

Partly, this comes from the fact that JAX's source code poses a number of unique challenges for Python's type annotation system. We'll outline them here.

### Challenge 1: pytype, mypy and developer friction
One challenge JAX faces is that package development must satisfy the constraints of two different static type checking systems, `pytype` (used by internal CI and internal Google projects) and `mypy` (used by external CI and external dependencies).

> Although the two type checkers have broad overlap in their behavior, each presents its own unique corner cases, as evidenced by the numerous `#type: ignore` and `#pytype: disable` statements througout the JAX codebase.

Reviewer: Smol typo:

```suggestion
Although the two type checkers have broad overlap in their behavior, each presents its own unique corner cases, as evidenced by the numerous `#type: ignore` and `#pytype: disable` statements throughout the JAX codebase.
```

This creates friction in development: internal contributors may iterate until tests pass, only to find that on export their pytype-approved code falls afoul of mypy.

> For external contributors, it's often the opposite: a recent example is {jax-issue}`#9596` which had to be rolled-back after it failed internal Google continous checks.

Reviewer: Smol typo:

```suggestion
For external contributors, it's often the opposite: a recent example is {jax-issue}`#9596` which had to be rolled-back after it failed internal Google continuous checks.
```

Each time we move a type annotation from Level 1 (`Any` everywhere) to Level 2 or 3 (stricter annotations), it introduces more of these potential sharp edges.

### Challenge 2: Array duck-typing

> One particular challenge for annotating JAX code is its heavy use of duck-typing. An input to a function marked `Array` in general could be one of many different types: a jax `DeviceArray`, a numpy `np.ndarray`, a numpy scalar, a Python scalar, a Python sequence, an object with an `__array__` attribute, an object with a `__jax_array__` attribute, or any flavor of `jax.Tracer`.

Reviewer: Nit: capitalization ("JAX", "NumPy")

```suggestion
One particular challenge for annotating JAX code is its heavy use of duck-typing. An input to a function marked `Array` in general could be one of many different types: a JAX `DeviceArray`, a NumPy `np.ndarray`, a NumPy scalar, a Python scalar, a Python sequence, an object with an `__array__` attribute, an object with a `__jax_array__` attribute, or any flavor of `jax.Tracer`.
```

2. In order to not add undue development friction (due to the internal/external CI differences), we would like to be conservative in the type annotation constructs we use: in particular, when it comes to recently-introduced mechanisms such as `ParamSpec` (PEP [PEP 612](https://peps.python.org/pep-0612/),), we would like to wait until support in mypy and other tools stabilizes before relying on them.
  One impact of this is that for the time being, when functions are decorated by jax transformations like `jit`, `vmap`, `grad`, etc. JAX will **strip all annotations**. This is because `ParamSpec` is still only partially supported; the PEP is slated for Python 3.10 (though it can be used before that via [typing-extensions](https://github.com/python/typing_extensions) and at the time of this writing mypy has a laundry-list of incompatibilities with the `ParamSpec`-based annotations (see [`ParamSpec` mypy bug tracker](https://github.com/python/mypy/issues?q=is%3Aissue+is%3Aopen++label%3Atopic-paramspec+)).
  We will revisit this question in the future once support for such features stabilizes.

> 3. JAX type annotations shoudl in general indicate the **intent** of APIs, rather than the implementation, so that the annotations become useful to communicate the contract of the API. This means that at times inputs that are valid at runtime may not be recognized as valid by the static type checker (a simple example is an arbitrary iterator passed in place of a shape in some function implementations).

Reviewer: Smol typo:

```suggestion
3. JAX type annotations should in general indicate the **intent** of APIs, rather than the implementation, so that the annotations become useful to communicate the contract of the API. This means that at times inputs that are valid at runtime may not be recognized as valid by the static type checker (a simple example is an arbitrary iterator passed in place of a shape in some function implementations).
```

  - `ArrayLike`
  - `DtypeLike`
  - `ShapeLike`
  - etc.
  
  Note that these will in general be simpler than the equivalent protocols used in {mod}`numpy.typing`. For example, in the case of `DtypeLike`, JAX does not support structured dtypes, so JAX can use a simpler implementation. Similarly, in `ArrayLike`, JAX generally does not support list or tuple inputs in most places, so the type definition will be simpler than the numpy analog.

5. Conversely, outputs of functions and methods should be typed as strictly as possible: for example, for a JAX function that returns an array, the output should be annotated with `jnp.ndarray` rather than `ArrayLike`. Functions returning a dtype should always be annotated `np.dtype`, and functions returning a shape should always be `Tuple[int]` or a strictly-typed NamedShape equivalent. For this purpose, we will implement in {mod}`jax.typing` several strictly-typed analogs of the permissive types mentioned above, namely:

  - `NDArray` (perhaps this could be equivalent to `jnp.ndarray`?)
  - `DType` (perhaps this could be simply `np.dtype`?)
  - `Shape`
  - `NamedShape`
  - etc.

> 6. Aside from common typing protocols gathered in `jax.typing`, we should err on the side of simplicity, and avoid constructing overly-complex protocols for arguments passed to API functions, and instead use simple unions such as `Union[simple_type, Any]` in the case that the full type specification of the API cannot be succinctly specified. This is a comprimise that achieves the goals of Level 1 and 2 annotations, while punting on Level 3 in favor of simplicity.

Reviewer: ```suggestion
6. Aside from common typing protocols gathered in `jax.typing`, we should err on the side of simplicity, and avoid constructing overly-complex protocols for arguments passed to API functions, and instead use simple unions such as `Union[simple_type, Any]` in the case that the full type specification of the API cannot be succinctly specified. This is a compromise that achieves the goals of Level 1 and 2 annotations, while punting on Level 3 in favor of simplicity.
```

# Type Annotation Roadmap for JAX

_jakevdp@_
_Aug 2022_

## Background

Python 3.0 introduced optional function annotations ([PEP 3107](https://peps.python.org/pep-3107/)), which were later codified for use in static type checking around the release of Python 3.5 ([PEP 484](https://peps.python.org/pep-0484/)).
To some degree, type annotations and static type checking have become an integral part of many Python development workflows, and to this end we have added annotations in a number of places throughout the JAX API.
The current state of type annotations in JAX is a bit patchwork, and efforts to add more have been hampered by more fundamental design questions.
This doc attempts to summarize those issues and generate a roadmap for the goals and non-goals of type annotations in JAX.

Why do we need such a roadmap? Better/more comprehensive type annotations are a frequent request from users, both internally and externally.

> In addition, we frequently receive pull requests from external users seeking to improve JAX's type annotations: it's not always clear to the JAX team member reviewing the code whether such contributions are beneficial, particularly when they introduce complex Protocols to address the challenges inherent to full-fledged annotation of JAX's use of Python.

Reviewer: Suggestion:
- Split into shorter simple sentences with `. And`, as it may be hard to comprehend (tired users, users who aren't very proficient in English).
- Add PR examples.

For example:

 ```suggestion
In addition, we frequently receive pull requests from external users (for example, [PR #10322](https://github.com/google/jax/pull/10322), [PR #9917](https://github.com/google/jax/pull/9917)) seeking to improve JAX's type annotations. And it's not always clear to the JAX team member reviewing the code whether such contributions are beneficial, particularly when they introduce complex Protocols to address the challenges inherent to full-fledged annotation of JAX's use of Python.
```

This document details JAX's goals and recommendations for type annotations within the package.

## Why type annotations?

There are a number of reasons that a Python project might wish to annotate their code-base; we'll summarize them in this document as Level 1, Level 2, and Level 3.

> ### Level 1: annotations as documentation

Reviewer: Here and across the doc. Forgot to mention this:

> "... capitalize only the first word in the title, the first word in a subheading after a colon, and any proper nouns or other terms that are always capitalized a certain way."

From https://developers.google.com/style/capitalization#capitalization-in-titles-and-headings

```suggestion
### Level 1: Annotations as documentation
```

> When originally introduced in [PEP 3107](https://peps.python.org/pep-3107/), type annotations were motivated partly by the ability to use them as concise, inline documentation of function parameter types and return types. JAX has long utilized annotations in this manner; an example is the common pattern of creating type names aliased to `Any`. An example can be found in `lax/slicing.py` [[source](http://google3/third_party/py/jax/_src/lax/slicing.py;l=48-54;rcl=446366018)]:

Reviewer: @jakevdp This g3 link won't "work" unless this is intentional 🙂

Reviewer: This link may work for _everyone_: https://github.com/google/jax/blob/main/jax/_src/lax/slicing.py 🙂

```python
Array = Any
Shape = core.Shape

def slice(operand: Array, start_indices: Sequence[int],
          limit_indices: Sequence[int],
          strides: Optional[Sequence[int]] = None) -> Array:
  ...
```

> For the purposes of static type checking, this use of `Array = Any` for array annotations puts no constraint on the argument values (`Any` is equivalent to no annotation at all) but it does serve as a form of useful in-code documentation for the developer.

Reviewer: Suggestion:
- Split a longer sentence into two shorter ones to reduce cognitive overhead.

```suggestion
For the purposes of static type checking, this use of `Array = Any` for array annotations puts no constraint on the argument values (`Any` is equivalent to no annotation at all). But it does serve as a form of useful in-code documentation for the developer.
```

For the sake of generated documentation, the name of the alias gets lost (the [HTML docs](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.slice.html) for `jax.lax.slice` report operand as type `Any`), so the documentation benefit does not go beyond the source code.

A benefit of this level of type annotation is that it is never wrong to annotate a value with `Any`, so it will provide a concrete benefit to developers and users in the form of documentation, without added complexity of satisfying the stricter needs of any particular static type checker.

> ### Level 2: annotations for intelligent autocomplete

Reviewer: Suggestion: Add a 👾 

```suggestion
### Level 2: annotations for intelligent autocomplete

```

Many modern IDEs take advantage of type annotations to improve [intelligent code completion](https://en.wikipedia.org/wiki/Intelligent_code_completion) in IDEs. One example of this is the [PyLance](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance) extension for VSCode, which uses Microsofts' [pyright](https://github.com/microsoft/pyright) static type checker as a source of information for VSCode's [IntelliSense](https://code.visualstudio.com/docs/editor/intellisense) completions.

This use of type checking requires going further than the simple aliases used above; for example, knowing that the `slice` function returns an alias of `Any` named `Array` does not add any useful information to the code completion engine. However, were we to annotate the function with a `DeviceArray` return type, the autocomplete would know how to populate the namespace of the result, and thus be able to suggest more relevant autocompletions during the course of development.

> JAX has begun to add this level of type annotation in a few places; one example is the `jnp.ndarray` return type within the `jax.random` package [[source](http://google3/third_party/py/jax/_src/random.py;l=359;rcl=445522278)]:

Reviewer: 🤷  This link may not "work". Maybe you meant this https://github.com/google/jax/blob/main/jax/_src/random.py#L359

```python
def shuffle(key: KeyArray, x: Array, axis: int = 0) -> jnp.ndarray:
  ...
```

> In this case `jnp.ndarray` is an abstract base class that forward-declares the attributes and methods of JAX arrays ([see source](http://google3/third_party/py/jax/_src/numpy/ndarray.py;l=40;rcl=431999528)), and so PyLance in VSCode can offer the full set of autocompletions on results from this function. Here is the result in VSCode:

Reviewer: As before. Maybe this will work better https://github.com/google/jax/blob/main/jax/_src/numpy/ndarray.py#L41

Gradual typing of an existing duck-typed API means that the current annotation is implicitly `Any`, so changing this to a stricter type may present to users as a breaking change.

Broadly speaking, annotating intent better serves Level 1 type checking, while annotating implementation better serves Level 3, while Level 2 is more of a mixed bag (both intent and implementation are important when it comes to annotations in IDEs).

## JAX type annotation roadmap

With this framing (Level 1/2/3) and JAX-specific challenges in mind, we can begin to develop our roadmap for implementing consistent type annotations across the JAX project.
For JAX type annotation, we will be guided by the following principles:

### Goal of type annotations

We would like to support full, *Level 1, 2, and 3* type annotation as far as possible. In particular, this means that we should have restrictive type annotations on both inputs and outputs to public API functions.

### Avoid unstable typing mechanisms

> In order to not add undue development friction (due to the internal/external CI differences), we would like to be conservative in the type annotation constructs we use: in particular, when it comes to recently-introduced mechanisms such as `ParamSpec` (PEP [PEP 612](https://peps.python.org/pep-0612/),), we would like to wait until support in mypy and other tools stabilizes before relying on them.

Reviewer: Suggestion:
- The link here may be rendering like `(PEP PEP 612),)`.

```suggestion
In order to not add undue development friction (due to the internal/external CI differences), we would like to be conservative in the type annotation constructs we use: in particular, when it comes to recently-introduced mechanisms such as `ParamSpec` ([PEP 612](https://peps.python.org/pep-0612/)), we would like to wait until support in mypy and other tools stabilizes before relying on them.
```

There are a number of reasons that a Python project might wish to annotate their code-base; we'll summarize them in this document as Level 1, Level 2, and Level 3.

### Level 1: Annotations as documentation

When originally introduced in [PEP 3107](https://peps.python.org/pep-3107/), type annotations were motivated partly by the ability to use them as concise, inline documentation of function parameter types and return types. JAX has long utilized annotations in this manner; an example is the common pattern of creating type names aliased to `Any`. An example can be found in `lax/slicing.py` [[source](https://github.com/google/jax/blob/2bc3e39cd9104071ee39dacac22abd51b94eb27e/jax/_src/lax/slicing.py#L47-L58)]:

```python
Array = Any
Shape = core.Shape

def slice(operand: Array, start_indices: Sequence[int],
          limit_indices: Sequence[int],
          strides: Optional[Sequence[int]] = None) -> Array:
  ...
```

For the purposes of static type checking, this use of `Array = Any` for array annotations puts no constraint on the argument values (`Any` is equivalent to no annotation at all), but it does serve as a form of useful in-code documentation for the developer.

> For the sake of generated documentation, the name of the alias gets lost (the [HTML docs](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.slice.html) for `jax.lax.slice` report operand as type `Any`), so the documentation benefit does not go beyond the source code.

Reviewer 2: FYI the autodoc extension has [some control](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_type_aliases) over this.

Another trick we [use in Haiku](https://github.com/deepmind/dm-haiku/blob/78c2984969d5b360c19a12bb7eb609712b8312f8/docs/conf.py#L36) docs is adding `from __future__ import annotations` to each .py source when building docs. Not pretty, but gets the job done :)

Author: Thanks! I added this to the discussion.

On the other hand, this strictness can make the type checking process very brittle in packages that often rely on duck-typing rather than strict type-safe APIs.
You'll currently find code comments like `#type: ignore` (for mypy) or `#pytype: disable` (for pytype) peppered throughout the JAX codebase in several hundred places.
These typically represent cases where typing problems have arisen; they may be inaccuracies in JAX type annotations, or inaccuracies in the static type checker's ability to correctly follow the control flow in the code.
On occasion, they are due to real & subtle bugs in the behavior of pytype or mypy.
In rare cases, they may be due to the fact that JAX uses Python patterns that are difficult or even impossible to express in terms of Python's static type annotation syntax.

## Type annotation challenges for JAX

JAX currently has type annotations that are a mixture of different styles, and aimed at all three levels of type annotation discussed above.
Partly, this comes from the fact that JAX's source code poses a number of unique challenges for Python's type annotation system. We'll outline them here.

> ### Challenge 1: pytype, mypy and developer friction

Reviewer 2: I think the friction is actually a result of Python not having a precise description of what the type system is and how it is supposed to work. So, mypy, pytype, pyright, pyre, IntelliJ etc all implement a slightly different type system.

@kmillikin did a talk about this on the recent Typing Summit: https://youtu.be/BNTkWQfqP_c?t=8466

Author: Thanks for the link! That's a good point – but I think JAX faces unique friction here because we use two different type-checkers in our CI, and different populations of JAX developers have access to only one or the other while writing their code.

One challenge JAX currently faces is that package development must satisfy the constraints of two different static type checking systems, `pytype` (used by internal CI and internal Google projects) and `mypy` (used by external CI and external dependencies).
Although the two type checkers have broad overlap in their behavior, each presents its own unique corner cases, as evidenced by the numerous `#type: ignore` and `#pytype: disable` statements throughout the JAX codebase.

This creates friction in development: internal contributors may iterate until tests pass, only to find that on export their pytype-approved code falls afoul of mypy.
For external contributors, it's often the opposite: a recent example is {jax-issue}`#9596` which had to be rolled-back after it failed internal Google continuous checks.
Each time we move a type annotation from Level 1 (`Any` everywhere) to Level 2 or 3 (stricter annotations), it introduces more potential for such frustrating developer experiences.

> ### Challenge 2: array duck-typing

Reviewer 2: Do you think it might be possible to reduce the degree polymorphism at least in JAX internals? Public APIs could continue to use these highly polymorphic types, but the internals would only accept `DeviceArray`s.

Author: That's mostly the case already.

The advantage of #1, annotating intent, is that the annotations are more useful to the user in communicating the API contract; while for the developer the flexibility leaves room for refactoring when necessary. The down-side (particularly for gradually-typed APIs like JAX's) is that it's quite likely that user code exists which runs correctly, but would be flagged as incorrect by a type checker.
Gradual typing of an existing duck-typed API means that the current annotation is implicitly `Any`, so changing this to a stricter type may present to users as a breaking change.

Broadly speaking, annotating intent better serves Level 1 type checking, while annotating implementation better serves Level 3, while Level 2 is more of a mixed bag (both intent and implementation are important when it comes to annotations in IDEs).

## JAX type annotation roadmap

> With this framing (Level 1/2/3) and JAX-specific challenges in mind, we can begin to develop our roadmap for implementing consistent type annotations across the JAX project.

Reviewer 2: Will the roadmap/further details be published as a separate JEP?

Author: This is intended to be the roadmap.

For JAX type annotation, we will be guided by the following principles:

### Purpose of type annotations

> We would like to support full, *Level 1, 2, and 3* type annotation as far as possible. In particular, this means that we should have restrictive type annotations on both inputs and outputs to public API functions.

Reviewer 3: > we should have restrictive type annotations on both inputs and outputs

What does "restrictive" mean here? Does it conflict with the section header "inputs should be permissively typed" below?

> to public API functions

I take this to mean that we're not directly concerned with annotating private stuff, for its own sake, as far as this JEP is concerned. Is that right? If so, I like that and encourage it be highlighted more prominently!

Reviewer 2: > If so, I like that and encourage it be highlighted more prominently!

Ideally, with some rationale for why type annotations are less useful for JAX internals :)

Author: "restrictive" here means a type that would disallow some inputs. `x: Any` is not restrictive. `x: int` is. When I use "permissive" below, it's referring to the same dimension: we want inputs to be more restrictive than `Any` but still relatively permissive (e.g. `x: Union[int, float, Array]` instead of `x: Any` or `x:Array`) and we want outputs to be as restrictive as possible (e.g. `x: Array` rather than `x: Union[int, float, Array]`)

Author: > I take this to mean that we're not directly concerned with annotating private stuff, for its own sake, as far as this JEP is concerned. Is that right? If so, I like that and encourage it be highlighted more prominently!

I think internal annotations are useful as well, and my intent was not to recommend against them in this JEP.

> We would like to support full, *Level 1, 2, and 3* type annotation as far as possible. In particular, this means that we should have restrictive type annotations on both inputs and outputs to public API functions.

Reviewer 3: > to support full, *Level 1, 2, and 3* type annotation as far as possible

From reading the rest of this note, and the suggestive ordering of those numbers, it seems like you have some degree of prioritization/preference in mind (e.g. 1 typically more important than 3). Is that the case?

Author: It's not an order of preference, it's an order of how complicated/specific the types must be.

Author: (my own opinion is that Level 3 is usually more trouble than its worth, especially for gradual typing of established packages where APIs have developed in a free-form duck-typing world typical of historical Python practices, but our loudest users seem pretty intent on Level 3 types wherever possible)

Reviewer 3: Nice – that's clarifying. I wasn't sure what the level numbers were meant to convey exactly.

> my own opinion is that comprehensive Level 3 is usually more trouble than its worth [...] but [...] users seem pretty intent on Level 3 types wherever possible

Maybe state this? We want to have concrete choices to act on (as you say, this is a "roadmap"). I'd be supportive of something like "we should at least aim for 2, though note some want 3 and that can come next" – or whatever you'd actually propose instead. We're not going to have a fully annotated codebase tomorrow.

### Annotate for intent

JAX type annotations should in general indicate the **intent** of APIs, rather than the implementation, so that the annotations become useful to communicate the contract of the API. This means that at times inputs that are valid at runtime may not be recognized as valid by the static type checker (a simple example is an arbitrary iterator passed in place of a shape in some function implementations).

> ### Inputs should be permissively-typed

Reviewer 3: ```suggestion
### Inputs should be permissively typed
```

Inputs to JAX functions and methods should be typed as permissively as is reasonable: for example, while shapes are typically tuples, functions that accept a shape should accept arbitrary sequences. Similarly, functions that accept a dtype need not require an instance of class `np.dtype`, but rather any dtype-convertible object. This might include strings, built-in scalar types, or dtype-adjacent classes such as `np.float64` and `jnp.float64`. In order to make this as uniform as possible across the package, we will add a {mod}`jax.typing` module with common type specifications, starting with broad categories such as:

- `ArrayLike`
- `DtypeLike`
- `ShapeLike`
- etc.
  
Note that these will in general be simpler than the equivalent protocols used in {mod}`numpy.typing`. For example, in the case of `DtypeLike`, JAX does not support structured dtypes, so JAX can use a simpler implementation. Similarly, in `ArrayLike`, JAX generally does not support list or tuple inputs in most places, so the type definition will be simpler than the NumPy analog.

> ### Outputs should be strictly-typed

Reviewer 3: ```suggestion
### Outputs should be strictly typed
```

> Conversely, outputs of functions and methods should be typed as strictly as possible: for example, for a JAX function that returns an array, the output should be annotated with `jnp.ndarray` rather than `ArrayLike`. Functions returning a dtype should always be annotated `np.dtype`, and functions returning a shape should always be `Tuple[int]` or a strictly-typed NamedShape equivalent. For this purpose, we will implement in {mod}`jax.typing` several strictly-typed analogs of the permissive types mentioned above, namely:

Reviewer 4: `Tuple[int, ...]`?

There are a number of reasons that a Python project might wish to annotate their code-base; we'll summarize them in this document as Level 1, Level 2, and Level 3.

### Level 1: Annotations as documentation

> When originally introduced in [PEP 3107](https://peps.python.org/pep-3107/), type annotations were motivated partly by the ability to use them as concise, inline documentation of function parameter types and return types. JAX has long utilized annotations in this manner; an example is the common pattern of creating type names aliased to `Any`. An example can be found in `lax/slicing.py` [[source](hhttps://github.com/google/jax/blob/2bc3e39cd9104071ee39dacac22abd51b94eb27e/jax/_src/lax/slicing.py#L47-L58)]:

Reviewer: Smol typo in `https` @jakevdp
